Kernel/Arch/x86/common/Boot/boot.S

```diff
@@ -390,10 +390,10 @@ pae_supported:
     movl $(boot_pdpt - KERNEL_BASE), %edi
 #if ARCH(X86_64)
     movl $((boot_pd0 - KERNEL_BASE) + 3), 0(%edi)
-    movl $((boot_pd3 - KERNEL_BASE) + 3), 24(%edi)
+    movl $((boot_pd3 - KERNEL_BASE) + 3), (8 * (KERNEL_BASE >> 30))(%edi)
 #else
     movl $((boot_pd0 - KERNEL_BASE) + 1), 0(%edi)
-    movl $((boot_pd3 - KERNEL_BASE) + 1), 24(%edi)
+    movl $((boot_pd3 - KERNEL_BASE) + 1), (8 * (KERNEL_BASE >> 30))(%edi)
```

Comment:
you should &0x1FF this value, not that you take the other possible upper bits with you

Reply:
Which value? `$((boot_pd3 - KERNEL_BASE) + 1)`? To be honest, I don't really tried to understand deeply this piece of code (maybe I should!), so if you could explain the problem you see here in a bit more detail, that'd be great :)

Reply:
I mean the KERNEL_BASE >> 30
The page decomposition looks something like 
```
PAGED ADDRESS LAYOUT x86_64 | PAGED ADDRESS LAYOUT x86
       16             9     |9(2 in 32)    9         9          12
0b1111111111111111 111111111|111111111 111111111 111111111 111111111111
 | sign extension | pml4 ind| pdpe ind| pde ind | pte ind | page offset
```
So in 64 bit, bits higher than the 31th (32 bit max) and even higher than the 38th bit (end of the table layer) might be set in the address, which would lead to a higher offset than desired,
@gunnarbeutner  for confirmation of the thought process

Reply:
Yeah, I was lazy and/or just missed this. Good catch. :+1:

Reply:
FWIW the next patch series gets rid of this. Also, the kernel wouldn't boot with a KERNEL_BASE > 512GB anyway. :)